File: 백준/레벨Gold IV/1976. 여행 가자/여행 가자.ts
// @ts-ignore
import fs = require('fs');

const read = fs.readFileSync(0, 'utf-8');
const lines = read.split('\n').map((line: string) => line.split(' ').map(Number));

const disjointSet = () => {
  // 부모 노드 초기화
  const [N, M] = lines[0];
  const parent = Array.from({ length: N + 1 }, (_, i) => i);

  const find = (x: number): number => {
    const nodes: number[] = [];
    while (parent[x] !== x) {
      nodes.push(x);
      x = parent[x];
    }
    // 한번에 경로 압축
    for (const node of nodes) {
      parent[node] = x;
    }
    return x;
  };

  const union = (a: number, b: number): void => {
    const [rootA, rootB] = [find(a), find(b)];
    if (rootA === rootB) return;

    if (rootA > rootB) parent[rootA] = rootB;
    else parent[rootB] = rootA;
  };

  // 분리 집합
  for (let i = 1; i <= N; i++) {
    const line = lines[i + 1];

    for (let j = 1; j <= N; j++) {
      if (line[j - 1] === 1) {
        union(i, j);
      }
    }
  }

  // 경로 검증
  const cities = lines[N + 2];
  const root = find(cities[0]);
  const isPossible = cities.every((city: number) => find(city) === root);
  console.log(isPossible ? 'YES' : 'NO');
};

disjointSet();

File: 백준/레벨Gold IV/1976. 여행 가자/여행 가자.py
from sys import stdin

input = stdin.readline

def disjoint_set():
    # 부모 노드 초기화
    N = int(input().rstrip())
    M = int(input().rstrip())
    parent = list(range(N+1))

    def find(x):
        nodes = []
        while parent[x] != x:
            nodes.append(x)            
            x = parent[x]
        # 한번에 경로 압축
        for node in nodes:
            parent[node] = x
        return x

    def union(a, b):
        root_a, root_b = find(a), find(b)
        if root_a == root_b:
            return

        if root_a > root_b:
            parent[root_a] = root_b
        else:
            parent[root_b] = root_a
    
    # 분리 집합
    for i in range(N):
        line = list(map(int, input().split()))

        for j in range(N):
            if line[j] == 1:
                union(i+1, j+1)

    # 경로 검증
    cities = list(map(int, input().split()))
    root = find(cities[0])
    isPossible = all(find(city) == root for city in cities)
    print("YES" if isPossible else "NO")

disjoint_set()